### ISSUE TITLE
Tileset editor breaks if you open and close it enough times (3.4beta regression)

### ISSUE THREAD

**Issue by ShatReal** [godotengine/godot#52760]
### Godot version

3.4-beta4

### System information

Windows 10

### Issue description

See title. By breaks, I mean that the grid is gone and it's impossible to select tiles or add new ones. And by times, I mean like three.

### Steps to reproduce

1. Make a TileSet
2. Open and close the editor until it breaks

**Comment by boruok** [comment]
Tileset editor buttons becomes non responsible(couldn't add collision, autotile, navigation...) grid becomes invisible. **Until you zoom in or zoom out**... can you confirm that?

**Comment by ShatReal** [comment]
Oh, zooming in and zooming out fixed it. Yes, I can confirm that.

**Comment by Calinou** [comment]
@boruok @ShatReal Can you reproduce this in 3.3.3?

**Comment by ShatReal** [comment]
No, I can't.

**Comment by Calinou** [comment]
If you can [compile the engine from source](https://docs.godotengine.org/en/stable/development/compiling/index.html), you could look into [bisecting the regression](https://docs.godotengine.org/en/stable/community/contributing/bisecting_regressions.html) to greatly speed up troubleshooting.

**Comment by akien-mga** [comment]
> If you can compile the engine from source, you could look into bisecting the regression to greatly speed up troubleshooting.

You can test 3.4 beta 1 through beta 4 first to narrow it down.

I've tried on 3.4 beta 4 but I don't understand what the bug should be exactly, the description is a bit short. Could you add screenshots or a video that demonstrate where the bug happens and how it looks like?

**Comment by ShatReal** [comment]
The bug appeared in beta3, beta2, and beta1. So something in beta1 caused it.

**Comment by boruok** [comment]
alright folks, that was a stressful day for my potato machine, but looks like we found it.
bruteforce steps taken:
1. generated commit log with **git log 3.3.3-stable..3.x > log.log**
2. cut list into 2 chunks, select random commit, download repo at specifc commit with -> https://codeload.github.com/godotengine/godot/zip/hashcode, build it, if tileset editor works -> bug is upper, if not -> lower. Repeat until you find it.
3. found it by reverting this commit -> https://github.com/godotengine/godot/commit/c8868fcaab22fbe81d88a0afcaecf222cd362e0a

can anyone confirm that?

**Comment by Calinou** [comment]
cc @HaSa1002

**Comment by HaSa1002** [comment]
Using Zylanns editor debugger plugin, I may found what is happening (or isn't).
Working:
![grafik](https://user-images.githubusercontent.com/14185889/134068185-3a097926-a20e-4320-bcc6-411220e89eb1.png)
Non-working:
![grafik](https://user-images.githubusercontent.com/14185889/134067967-b7ca6d8d-a602-4dc8-ba51-6c699bb3007b.png)

I honestly don't think that my change is the root of the evil and rather a symptom. (We would see regression in way more cases)

I don't really know the TileSet editor code so it may take me time to figure things out. Before I do that, I will try to recreate the bug

**Comment by HaSa1002** [comment]
The node in question in code (`tile_set_editor_plugin.cpp`) is `workspace`. However, it is used in 194 cases. From the looks of it, the control node is updated manually, so I guess it doesn't call update after a size change, because of the previous behaviour

**Comment by akien-mga** [comment]
> alright folks, that was a stressful day for my potato machine, but looks like we found it.
> bruteforce steps taken:
> 
> 1. generated commit log with **git log 3.3.3-stable..3.x > log.log**
> 2. cut list into 2 chunks, select random commit, download repo at specifc commit with -> [codeload.github.com/godotengine/godot/zip/hashcode](https://codeload.github.com/godotengine/godot/zip/hashcode), build it, if tileset editor works -> bug is upper, if not -> lower. Repeat until you find it.
>3. found it by reverting this commit -> [c8868fc](https://github.com/godotengine/godot/commit/c8868fcaab22fbe81d88a0afcaecf222cd362e0a)

FYI, the steps you performed are called "bisecting", and there's a built-in Git tool to do exactly that which could have saved you some time :)
https://docs.godotengine.org/en/latest/community/contributing/bisecting_regressions.html

**Comment by follower** [comment]
A drive-by observation:

 * https://github.com/godotengine/godot/commit/c8868fcaab22fbe81d88a0afcaecf222cd362e0a changed the order of the `_size_changed();` call in relation to the `data.updating_last_minimum_size = false;` line: https://github.com/godotengine/godot/blob/489b49c6b89f68b9f3cc3f8a9579c6ed1e2fdf82/scene/gui/control.cpp#L290-L303

 * The value of `data.updating_last_minimum_size` is used here: https://github.com/godotengine/godot/blob/489b49c6b89f68b9f3cc3f8a9579c6ed1e2fdf82/scene/gui/control.cpp#L2494-L2521

So, perhaps try something like this instead (untested, manually edited):

```patch
diff --git a/scene/gui/control.cpp b/scene/gui/control.cpp
index bff3024e38e..76ceef27821 100644
--- a/scene/gui/control.cpp
+++ b/scene/gui/control.cpp
@@ -285,15 +285,11 @@ void Control::_update_minimum_size() {
 	}
 
 	Size2 minsize = get_combined_minimum_size();
-	if (minsize.x > data.size_cache.x ||
-			minsize.y > data.size_cache.y) {
+      if (minsize != data.last_minimum_size) {
 		_size_changed();
 }
 
 	data.updating_last_minimum_size = false;
 
 	if (minsize != data.last_minimum_size) {
 		data.last_minimum_size = minsize;
 		emit_signal(SceneStringNames::get_singleton()->minimum_size_changed);
 	}
 }

```

I'd be inclined to test this by printing out the `Control`s that are traversed before/after the minimum size change.

Although I note that I don't know what the relationship is between e.g. `data.size_cache.x` and `data.last_minimum_size` values are either.

**Comment by akien-mga** [comment]
Fixed by #55059.
